pkg/controller/master/backup/restore.go

```diff
@@ -513,54 +560,160 @@ func (h *RestoreHandler) updateOwnerRefAndTargetUID(vmRestore *harvesterv1.Virtu
+		volumeSnapshot, err := h.getOrCreateVolumeSnapshot(vmRestore, volumeBackup)
+		if err != nil {
+			return err
+		}
+		dataSourceName = volumeSnapshot.Name
+	}
+
+	annotations := map[string]string{}
+	for key, value := range volumeBackup.PersistentVolumeClaim.ObjectMeta.Annotations {
+		needSkip := false
+		for _, prefix := range restoreAnnotationsToDelete {
+			if strings.HasPrefix(key, prefix) {
+				needSkip = true
+				break
+			}
+		}
+		if !needSkip {
+			annotations[key] = value
+		}
+	}
+	annotations[restoreNameAnnotation] = vmRestore.Name
+
 	sourcePVC := volumeBackup.PersistentVolumeClaim
-	pvc := &corev1.PersistentVolumeClaim{
+	spec := *sourcePVC.Spec.DeepCopy()
+	spec.VolumeName = ""
+	spec.DataSource = &corev1.TypedLocalObjectReference{
+		APIGroup: pointer.StringPtr(snapshotv1.SchemeGroupVersion.Group),
+		Kind:     volumeSnapshotKindName,
+		Name:     dataSourceName,
+	}
+
+	_, err := h.pvcClient.Create(&corev1.PersistentVolumeClaim{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:        volumeRestore.PersistentVolumeClaim.ObjectMeta.Name,
 			Namespace:   vmRestore.Namespace,
 			Labels:      sourcePVC.ObjectMeta.Labels,
-			Annotations: sourcePVC.ObjectMeta.Annotations,
-		},
-		Spec: *sourcePVC.Spec.DeepCopy(),
-	}
-
-	pvc.SetOwnerReferences([]metav1.OwnerReference{
-		{
-			APIVersion:         harvesterv1.SchemeGroupVersion.String(),
-			Kind:               vmRestoreKindName,
-			Name:               vmRestore.Name,
-			UID:                vmRestore.UID,
-			Controller:         pointer.BoolPtr(true),
-			BlockOwnerDeletion: pointer.BoolPtr(true),
+			Annotations: annotations,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion:         harvesterv1.SchemeGroupVersion.String(),
+					Kind:               vmRestoreKindName,
+					Name:               vmRestore.Name,
+					UID:                vmRestore.UID,
+					Controller:         pointer.BoolPtr(true),
+					BlockOwnerDeletion: pointer.BoolPtr(true),
+				},
+			},
 		},
+		Spec: spec,
 	})
-	if volumeBackup.Name == nil {
-		return fmt.Errorf("missing VolumeSnapshot name")
+	return err
+}
+
+func (h *RestoreHandler) getOrCreateVolumeSnapshotContent(
+	vmRestore *harvesterv1.VirtualMachineRestore,
+	volumeBackup harvesterv1.VolumeBackup,
+) (*snapshotv1.VolumeSnapshotContent, error) {
+	volumeSnapshotContentName := h.constructVolumeSnapshotContentName(vmRestore.Namespace, vmRestore.Name, *volumeBackup.Name)
+	if volumeSnapshotContent, err := h.snapshotContentCache.Get(volumeSnapshotContentName); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return nil, err
+		}
+	} else {
+		return volumeSnapshotContent, nil
 	}
 
-	if pvc.Annotations == nil {
-		pvc.Annotations = make(map[string]string)
+	lhBackup, err := h.lhbackupCache.Get(util.LonghornSystemNamespaceName, *volumeBackup.LonghornBackupName)
+	if err != nil {
+		return nil, err
 	}
+	snapshotHandle := fmt.Sprintf("bs://%s/%s", volumeBackup.PersistentVolumeClaim.ObjectMeta.Name, lhBackup.Name)
```

Comment:
NIT: Could we add a reference for this magic URL?

```suggestion
	// Ref: https://longhorn.io/docs/1.2.3/snapshots-and-backups/csi-snapshot-support/restore-a-backup-via-csi/#restore-a-backup-that-has-no-associated-volumesnapshot
	snapshotHandle := fmt.Sprintf("bs://%s/%s", volumeBackup.PersistentVolumeClaim.ObjectMeta.Name, lhBackup.Name)
```

Reply:
Added it. Thank you.